packages/core/src/components/panel-stack/panelStack.tsx

```diff
@@ -0,0 +1,142 @@
+
+/** Add this interface to your panel components props type. */
+export interface IPanelProps {
+    /**
+     * Call this method to close the current panel. It should be used sparingly as there is always a back button displayed that will call
+     * this method. If there is only one panel on the stack, it cannot be closed and this method will do nothing.
+     */
+    closePanel(): void;
+    /**
+     * Call this method to open a new panel on the top of the stack.
+     * @param component The React component of the new panel.
+     * @param props The props to be passed to the new panel.
+     * @param options Additional options for the new panel.
+     */
+    openPanel<P>(component: React.ComponentClass<P & IPanelProps>, props: P, options: IPanelOptions): void;
+}
+
+export interface IPanel<P = {}> extends IPanelOptions {
+    /** The component being rendered by this panel. */
+    component: React.ComponentClass<P & IPanelProps>;
+    /** The props passed directly to the component prop. */
+    props?: P;
+}
+
+export interface IPanelOptions {
+    /** The title to be displayed above this panel. It is also used as the back button for any panel opened by this panel. */
+    title: string;
+}
+
+export interface IPanelStackProps extends IProps {
+    /** The panel to show when initially rendering the component */
+    initialPanel: IPanel;
+    /**
+     * Callback invoked when the user presses the back button or a panel wants to close the visible panel.
+     * @param oldPanel The panel that was just removed from the stack.
+     */
+    onClose?(oldPanel: IPanel): void;
+    /**
+     * Callback invoked when a panel wants to open a new panel.
+     * @param newPanel The panel that was just added to the stack.
+     */
+    onOpen?(newPanel: IPanel): void;
+}
+
+export interface IPanelStackState {
+    /** Whether the stack is currently animating the push or pop of a panel. */
+    direction: "push" | "pop";
+    /** The current stack of panels. The first panel in the stack will be displayed. */
+    stack: IPanel[];
+}
+
+export class PanelStack extends React.PureComponent<IPanelStackProps, IPanelStackState> {
+    public constructor(props: IPanelStackProps, context?: any) {
+        super(props, context);
+        this.state = {
+            direction: "push",
+            stack: [props.initialPanel],
+        };
+    }
+
+    public render() {
+        const className = classNames(Classes.PANELSTACK, this.props.className);
+        return (
+            <TransitionGroup className={className} component="div">
+                {this.renderCurrentPanel()}
+            </TransitionGroup>
+        );
+    }
+
+    private renderCurrentPanel() {
+        const { stack } = this.state;
+        const [activePanel, previousPanel] = stack;
+        return (
+            <CSSTransition
+                classNames={`${Classes.PANELSTACK_TRANSITION}-${this.state.direction}`}
+                key={stack.length}
```

Comment:
this does not feel safe to use as a key. the size of the stack is not a unique identifier of the panel. not really sure what to do tho, needs investigation.